src/libraries/System.Linq/src/System/Linq/Chunk.cs

```diff
@@ -93,18 +128,22 @@ private static IEnumerable<TSource[]> ChunkIterator<TSource>(IEnumerable<TSource
                         }
                     }
 
-                    // Extract the chunk array to yield, then clear out any references in our storage so that we don't keep
-                    // objects alive longer than needed (the caller might clear out elements from the yielded array.)
-                    var chunk = new TSource[i];
-                    Array.Copy(array, 0, chunk, 0, i);
-                    if (RuntimeHelpers.IsReferenceOrContainsReferences<TSource>())
+                    // If the array has the right size return it directly
+                    if (i == array.Length)
+                    {
+                        yield return array;
+                    }
+                    else
                     {
-                        Array.Clear(array, 0, i);
+                        // Otherwise, copy to a correctly sized array and null out the original array so that we don't keep
+                        // objects alive longer than needed (the caller might clear out elements from the yielded array.)
+                        var chunk = new TSource[i];
+                        Array.Copy(array, 0, chunk, 0, i);
+                        array = null;
```

Comment:
With your changes, is the array still being lifted to a field on the display class?  I wouldn't expect it to be.  If it's not, then it's not necessary to null out the array, and if it is, then the array would also need to be nulled out in the i == array.Length case.  Assuming it's not being lifted, this path should just use Array.Resize, e.g.
```C#
if (i != array.Length)
{
    Array.Reset(ref array, i);
}

yield return array;
```